### PR TITLE
add support for bigint

### PR DESCRIPTION
**Pull request by maximelkin**
todo: fix tests

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -20,6 +20,9 @@ module.exports = {
     'no-fallthrough': warning,
     'require-atomic-updates': 0,
   },
+  globals: {
+    BigInt: true,
+  },
```
Comment (briandamaged) [inline .eslintrc.js]: Do we need to officially drop support for node 8 first?

File: test/integration/builder/inserts.js [hunk 2]
```diff
@@ -1302,5 +1302,17 @@ module.exports = function(knex) {
           });
       });
     });
+
+    describe('#3553 insert with bigint', function() {
+      if (typeof BigInt === 'undefined') return;
```
Comment (briandamaged) [inline test/integration/builder/inserts.js]: Hmm... does `"#3553 insert with bigint"` still get printed in the test case output?  If so, then we might want to move the conditional outside of the `describe(..)` block:

```
if(typeof BigInt !== 'undefined') {
  describe('#3553 insert with bigint', function() {
    //
  });
}
```

File: test/integration/builder/inserts.js [hunk 3]
```diff
@@ -1302,5 +1302,17 @@ module.exports = function(knex) {
...
+
+      it('should allow insert with BigInt', async function() {
+        await knex('accounts').insert([{ id: BigInt(12) }]);
```
Comment (briandamaged) [inline test/integration/builder/inserts.js]: Might want to:

1. Test w/ a number that is really large.  (Ex:  `BigInt("1234567812345678123456781234567812345678")`)
2. Read the number back out of the database as well.

Also: it looks like the `MSSQL` changes the datatype depending upon the size of the integer.  So, there are at least 5 different scenarios that should be tested there.  (And, probably more like 7-9 cases if you count the boundary conditions as well).

File: test/integration/builder/inserts.js [hunk 4]
```diff
@@ -1302,5 +1302,17 @@ module.exports = function(knex) {
...
+      });
+      it('should allow insert array with BigInt', async function() {
+        if (knex.client.driverName !== 'pg') return true;
+        await knex('accounts').insert([{ id: [BigInt(12)] }]);
```
Comment (briandamaged) [inline test/integration/builder/inserts.js]: It would probably be a good idea to use a larger number here as well.

File: test/integration/builder/inserts.js [hunk 5]
```diff
@@ -1302,5 +1302,17 @@ module.exports = function(knex) {
...
     });
+
+    describe('#3553 insert with bigint', function() {
+      if (typeof BigInt === 'undefined') return;
+
+      it('should allow insert with BigInt', async function() {
+        await knex('accounts').insert([{ id: BigInt(12) }]);
+      });
+      it('should allow insert array with BigInt', async function() {
+        if (knex.client.driverName !== 'pg') return true;
```
Comment (briandamaged) [inline test/integration/builder/inserts.js]: Might want to do `return this.skip()` for now.

(Eventually, we should figure out a way to restructure the test cases so that these skipped scenarios get omitted.  But, marking them as `pending` will at least remind us to revisit this later)

File: test/integration/builder/updates.js [hunk 6]
```diff
@@ -339,5 +339,36 @@ module.exports = function(knex) {
...
     });
+
+    describe('#3553 update with bigint', function() {
+      if (typeof BigInt === 'undefined') return;
+
+      it('should allow insert with BigInt', async function() {
+        const firstName = '#3553';
+        await knex('accounts').insert([{ balance: 5, first_name: firstName }]);
+        await knex('accounts').update([
+          { balance: BigInt(Number.MAX_SAFE_INTEGER) },
```
Comment (briandamaged) [inline test/integration/builder/updates.js]: `Number.MAX_SAFE_INTEGER` is still a regular integer, correct?  If so, then you might want to use a larger number here.

Comment (briandamaged) [reply test/integration/builder/updates.js]: Oh, and of course: read the number back out of the database as well.  (Just to make sure it didn't get truncated or anything like that)

File: test/integration/datatype/bigint.js [hunk 7]
```diff
@@ -122,4 +122,33 @@ module.exports = function(knex) {
...
+      })
+      .then(function() {
+        return knex(tableName).insert({ value: value });
+      })
+      .then(function() {
+        return knex(tableName).first('value');
+      })
+      .then(function(response) {
+        expect(response.value).to.be.eql(value.toString());
+      });
```
Comment (briandamaged) [inline test/integration/datatype/bigint.js]: You could use `async` / `await` syntax here if you wanted to.

File: test/unit/query/builder.js [hunk 8]
```diff
@@ -132,7 +132,7 @@ function testquery(chain, valuesToCheck, selectedClients) {
     newChain.client = selectedClients[key];
     const sqlString = newChain.toQuery();
     const checkValue = valuesToCheck[key];
-    expect(checkValue).to.equal(sqlString);
+    expect(checkValue).to.equal(sqlString, `for driver ${key}`);
```
Comment (briandamaged) [inline test/unit/query/builder.js]: Unrelated enhancement?

File: test/unit/query/builder.js [hunk 9]
```diff
@@ -9863,4 +9863,31 @@ describe('QueryBuilder', () => {
       }
     );
   });
+
+  it('#3553 should correctly compile insert with BigInt', function() {
+    testquery(
+      qb()
+        .insert([{ a: BigInt(12) }])
```
Comment (briandamaged) [inline test/unit/query/builder.js]: `BigInt("1234567182348761238746512348765")` just to make sure it's not coercing the values back into regular `Number` instances.

**Comment by maximelkin** [comment]
@briandamaged Thanks for review! I will add test cases

File: test/integration/builder/inserts.js (hunk 2 above)
Comment (maximelkin) [reply test/integration/builder/inserts.js]: I think this really not matters, because with node 8 drop this line will be removed

**Comment by kibertoad** [comment]
Unless we absolutely want this in 0.20, we can mark it with `breaking change` label and drop Node 8 already.

**Comment by maximelkin** [comment]
@kibertoad this merge is not breaking node 8 support
but i think its ok to merge this in 1.x version
(sorry for long inactivity, with this merge)

**Comment by kibertoad** [comment]
@maximelkin Didn't it include specific checks for pre-10 Node, though?

**Comment by maximelkin** [comment]
typeof something === 'bigint'
this is compatible with node 8

File: .eslintrc.js (hunk 1 above)
Comment (maximelkin) [reply .eslintrc.js]: fixed, now globals for test/ directory only

**Comment by maximelkin** [comment]
todo: workaround for https://github.com/mapbox/node-sqlite3/issues/1058

**Comment by maximelkin** [comment]
When selecting decimal from postgres, type of resulting value can be string or number, depending on size of data
This is because of old version pg-types for https://github.com/brianc/node-postgres

**Comment by maximelkin** [comment]
mssql when selecting decimals without any settings give simple numbers, so this explodes

**Comment by maximelkin** [comment]
God bless mysql drivers - the most simple configuration for getting strings for decimals and bigints

**Comment by maximelkin** [comment]
On oracledb side probably we should always set fetchAsString and then parse to normal integer by passed metadata (numeric, int64, float64 -> string, less precision numbers to js native)

**Comment by maximelkin** [comment]
mysql give errors because accounts.id declared as bigint autoincrement

**Comment by maximelkin** [comment]
Maybe we should split merge request into fixing support of int64 and decimals for databases and support of js bigint?

**Comment by kibertoad** [comment]
@maximelkin Yeah, split makes sense.

**Comment by kibertoad** [comment]
@maximelkin Should we extract the working part and merge that for now?

**Comment by maximelkin** [comment]
I think this pr will be adding support for querybuilding with .toString() only, but not sure is this useful because some libs dont support bigint yet

File: .eslintrc.js (hunk 1 above)
Comment (kibertoad) [reply .eslintrc.js]: Support for Node 8 was dropped already, so this is good to go.

**Comment by kibertoad** [comment]
@maximelkin I would recommend pulling from master for this PR for vastly faster CI :)
